### DISCUSSION TITLE
Send buffer code to REPL

### DISCUSSION THREAD

**Discussion by zippeurfou** [mfussenegger/nvim-dap#962] [Q&A] [answered]
Hi there,
I would love to be able to manually send code I have in other buffer to the REPL buffer like plugin like iron.nvim does.
Context:
I use the repl to debug python code while at a breakpoint. It helps me several ways using:
```
map("n", "<leader>Dr", "<cmd>lua require'dap'.repl.toggle()<cr>", "[R]epl toggle")
```
One issue I have is that the REPL somehow does not allow me to:
1. paste yanked code into it.
2. Send directly selection to it

I mostly use nvim-dap-python with debugpy. Let me know if this is not the right place for it.

**Comment by LiadOz** [comment, marked as answer]
Here is a snippet that allows you to add lines to the current prompt buffer, create a file called dap-repl.lua in your ftplugin folder:
```lua
local function add_line(line)
  local lnum = vim.fn.line('$')
  line = string.gsub(line, "\n", "")
  vim.api.nvim_buf_set_lines(0, lnum - 1, lnum, true, {'dap> ' .. line})
  vim.cmd('startinsert!')
end

vim.keymap.set('n', 'p', function() add_line(vim.fn.getreg('"')) end, {buffer = 0})
```
this will allow you to use p in the dap repl buffer.
you could play with it to achieve more complex operations.

**Comment by zippeurfou** [comment]
Thanks!
It works but it does not do multiline (which makes sense since you're removing it). Would you know how this would be possible?
I am guessing you know how since the image in your repo shows multiline :) 
(https://github.com/LiadOz/nvim-dap-repl-highlights)
Once you answer this I will put this as resolved. Thanks again.

**Reply by LiadOz** [reply]
You will probably need to wait for this https://github.com/mfussenegger/nvim-dap/pull/773 before you could do multi line stuff with nvim-dap but you could use my fork to try and use it https://github.com/LiadOz/nvim-dap/tree/LiadOz/multiline-inputs
Making it work with multilines seems to be more tricky so I won't attempt to do it now, Maybe once multiline is merged I will look into it. If you manage to write it let me know...

Also note that you need to configure the is_multiline function to the adapter to use multilines, here is a snippet from my config:
```lua
dap.adapters.python = function(cb, config)
  local function is_multiline(text)
    if string.sub(text, -1) == ':' then
      return true
    elseif string.sub(text, -1) == '\\' then
      return true
    elseif select(2, string.gsub(text, '"""', "")) == 1 then
      return true
    end
    return false
  end
  if config.request == 'attach' then
    cb({
      type = 'server',
      host = '127.0.0.1',
      port = 1337,
      is_multiline = is_multiline,
    })
  else
    cb({
      type = 'executable',
      command = 'python',
      args = {'-m', 'debugpy.adapter'},
      is_multiline = is_multiline,
    })
  end
end
```

**Reply by zippeurfou** [reply]
Thanks for it! I will put it as resolved as it is not trivial but I appreciate the solution.

**Comment by mfussenegger** [comment]
With `getregion` in neovim 0.10 you could use something like this:

```lua
local dap = require("dap")
vim.keymap.set("x", "<leader>di", function()
  local lines = vim.fn.getregion(vim.fn.getpos("."), vim.fn.getpos("v"))
  dap.repl.open()
  dap.repl.execute(table.concat(lines, "\n"))
end)
```

**Reply by wilsonchen719** [reply]
Hi @mfussenegger ,
Thank you for the replies and create such fantastic package : )
Tried out the solution, it works perfect!
